goto-symex: use function and called_function fields [blocks: #3126]

tautschnig opened the pull request:
goto_programt::instructionst::instructiont::function is going to go away, use
the fields that are already available within goto-symex.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
goto-symex: use function and called_function fields

goto_programt::instructionst::instructiont::function is going to go away, use
the fields that are already available within goto-symex.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fb5bdcf63c7478f11d277c49eab6a33bb1b8b092).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97888387

kroening commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -85,8 +85,10 @@ void goto_symext::symex_goto(statet &state)
-      frame.loop_iterations[goto_programt::loop_id(*state.source.pc)].count;
+    const auto &loop_id =
+      goto_programt::loop_id(state.source.function, *state.source.pc);
+
```
goto_programt::loop_id returns a temporary, so I would not use a reference here.

tautschnig replied:
Done, and I've now added a second commit to make sure we use `goto_programt::loop_id` everywhere.

tautschnig committed:
Use goto_programt::loop_id

This ensures we build loop identifiers in the same way everywhere.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 91c66b4a37b494e5d1ae4de55471657d874d6acd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97925014

peterschrammel reviewed (approved):
Consider https://github.com/diffblue/cbmc/pull/3828#discussion_r249272154

tautschnig committed:
symex: Consistently use function_id

"function" is too ambiguous and "function_identifier" is too long, and mixing
them is even worse.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: c6bdc44d90849afd611d2d51dd0ddcbbaebf5304).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97968538
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 67507db23071aa8e7b0ed217d4e4dcaa8e88f67c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97969840
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

tautschnig commented:
@peterschrammel It's perfectly possible that the TG failure is genuine and caused by renaming the `function` member to `function_id`. Could you please take a look and possibly create a suitable bump?

Marking this as do-not-merge for now.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 92ec183868d59cb5e2a0fa8b598feb683287a419).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98122649
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj commented:
TG update: https://github.com/diffblue/test-gen/pull/2852

allredj commented:
@tautschnig could you please rebase this PR onto develop (or at least on top of https://github.com/diffblue/cbmc/pull/3876) so that I can compile with TG?

tautschnig commented:
@allredj Rebase done.

allredj commented:
@tautschnig thanks. TG bump passing.

allredj commented:
Shall I merge?

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: f33f64e76399820bb0dda394679f3e0bee6419cb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98458367
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

tautschnig commented:
@allredj Thanks a lot, was away from my desk - merge done. Please merge the corresponding TG update.
